runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedilyFusedExecutableStage.java

```diff
@@ -0,0 +1,218 @@
+  // consume a PCollection materialized by this subgraph and can be fused into it).
+  private static final Logger LOG = LoggerFactory.getLogger(GreedilyFusedExecutableStage.class);
+
+  /**
+   * Returns an {@link ExecutableStage} where the initial {@link PTransformNode PTransform} is a
+   * Read which executes within the environment (with no input elements or {@link PCollectionNode
+   * PCollection}).
+   *
+   * <p>Once primitive Read transforms are replaced by an Impulse -> ParDo expansion, this
+   * construction should be removed.
+   */
+  public static ExecutableStage forRootTransform(
+      QueryablePipeline pipeline, PTransformNode rootTransform) {
+    checkArgument(
+        PTransformTranslation.READ_TRANSFORM_URN.equals(
+            rootTransform.getTransform().getSpec().getUrn()),
+        "Can only create an %s rooted at a %s node, got URN %s",
+        ExecutableStage.class.getSimpleName(),
+        PTransformTranslation.READ_TRANSFORM_URN,
+        rootTransform.getTransform().getSpec().getUrn());
+    return new GreedilyFusedExecutableStage(
+        pipeline, null, Collections.singleton(rootTransform));
+  }
+
+  /**
+   * Returns an {@link ExecutableStage} where the initial {@link PTransformNode PTransform} is a
+   * Remote GRPC Port Read, reading elements from the materialized {@link PCollectionNode
+   * PCollection}.
+   *
+   * @param initialNodes the initial set of sibling transforms to fuse into this node. All of the
+   *     transforms must consume the {@code inputPCollection} on a per-element basis, and must all
+   *     be mutually compatible.
+   */
+  public static ExecutableStage forGrpcPortRead(
+      QueryablePipeline pipeline,
+      PCollectionNode inputPCollection,
+      Set<PTransformNode> initialNodes) {
+    return new GreedilyFusedExecutableStage(pipeline, inputPCollection, initialNodes);
+  }
+
+  private final QueryablePipeline pipeline;
+  private final Environment environment;
+
+  @Nullable private final PCollectionNode maybeInputPCollection;
+
+  private final Set<PCollectionNode> fusedCollections;
+  private final Set<PTransformNode> fusedTransforms;
+  private final Set<PCollectionNode> materializedPCollections;
+
+  private final Queue<PCollectionNode> fusionCandidates;
+
+  private GreedilyFusedExecutableStage(
+      QueryablePipeline pipeline,
+      @Nullable PCollectionNode inputPCollection,
+      Set<PTransformNode> initialNodes) {
+    checkArgument(
+        !initialNodes.isEmpty(),
+        "%s must contain at least one %s.",
+        GreedilyFusedExecutableStage.class.getSimpleName(),
+        PTransformNode.class.getSimpleName());
+    // Choose the environment from an arbitrary node. The initial nodes may not be empty for this
+    // subgraph to make any sense, there has to be at least one processor node
+    // (otherwise the stage is GRPCRead -> GRPCWrite, which doesn't do anything).
```

Comment:
`GRPC` -> `gRPC`

Other places you called them `GRPC Read` vs here you call it `GRPCRead`

Reply:
Done